app/workers/facilities/pssg_download.rb

```diff
@@ -56,19 +56,31 @@ def extract_polygon(rings)
     end
 
     def insert_or_update_band(facility)
-      if facility.drivetime_bands.exists?(vha_facility_id: @id, name: @band_name)
-        Rails.logger.info "PSSG Band not updated: Facility #{@id}. Band #{@attributes&.dig('Name')}"
-      else
+      drive_time_band = facility.drivetime_bands.find_by(vha_facility_id: @id, name: @band_name)
+
+      if drive_time_band.nil?
         drive_time_band = facility.drivetime_bands.new(vha_facility_id: @id, name: @band_name)
         drive_time_band.unit = 'minutes'
         drive_time_band.min = round_band(@attributes&.dig('FromBreak'))
         drive_time_band.max = round_band(@attributes&.dig('ToBreak'))
-        drive_time_band.name = @band_name
         drive_time_band.polygon = extract_polygon(@rings)
+        drive_time_band.vssc_extract_date = @vssc_extract_date
 
-        Rails.logger.info "PSSG Band successfully saved: #{@band_name}" # temporary logging
         drive_time_band.save
-        facility.save
+        # Rails.logger.info "PSSG Band successfully inserted: #{@band_name}" # temporary logging
+      elsif @vssc_extract_date > drive_time_band.try(:vssc_extract_date)
+        # rubocop:disable Rails/SkipsModelValidations
+        drive_time_band.update_columns(
+          min: round_band(@attributes&.dig('FromBreak')),
+          max: round_band(@attributes&.dig('ToBreak')),
+          polygon: extract_polygon(@rings),
+          vssc_extract_date: @vssc_extract_date,
+          updated_at: Time.zone.now
+        )
```

Comment:
just out of curiosity, why skip validations?

Reply:
We are skipping validations and callbacks because they take much longer than the update itself. For instance when running locally 30 records can be updated in around 3 seconds on average, but it takes more than 200 seconds when passing through all of the extra steps that Active Record includes.

The field `polygon` is a giant array of latitude and longitude points (a lot of bands have thousands of points). In Postgres these are stored as PostGIS geographic polygons that allow us to quickly compare them to individual lat/long points. We use the gem `rgeo-geojson` to represent these PostGIS objects in Active Model. There is a lot of overhead involved in converting and comparing between JSON and these objects. Even comparing an unchanged polygon to itself to see if the column needs to be updated is very slow. By instead just checking if our source has updated their polygon and replacing our cached polygon if so, we are able to greatly speed up this process. Additionally `extract_polygon` does will throw an error if the band is malformed, so we have already vetted the value.

I suppose we could update the rest of the record with validations and callbacks turned on, but that would me sending two separate update statements to Postgres, which is less than ideal.